docs/src/converting_zygoterules.md

```diff
@@ -0,0 +1,136 @@
+# Converting ZygoteRules.@adjoint to `rrule`s
+
+[ZygoteRules.jl](https://github.com/FluxML/ZygoteRules.jl) is a legacy package similar to ChainRules but supporting [Zygote.jl](https://github.com/FluxML/Zygote.jl) only.
+
+If you have some rules written with ZygoteRules it is a good idea to upgrade them to use ChainRules instead.
+Zygote will still be able to use them, but so will other AD systems,
+and you will get access to some more advanced features.
+(Which Zygote may or may not then ignore).
+
+## Example
+Consider the function
+```julia
+struct Foo
+    a::Float64,
+    b::Float64
+end
+
+f(x, y::Foo, z) = 2*x + y.a
+```
+
+### ZygoteRules
+```julia
+@adjoint function f(x, y::Foo, z)
+    f_pullback(Ω̄) = (2Ω̄, NamedTuple(;a=Ω̄, b=nothing), nothing)
+    return f(x, y, z), f_pullback
+end
+```
+
+### ChainRules
+```julia
+function rrule(::typeof(f), x, y::Foo, z)
+    f_pullback(Ω̄) = (NoFields(), 2Ω̄, Tangent{Foo}(;a=Ω̄), ZeroTangent())
+    return f(x, y, z), f_pullback
+end
+```
+
+## Write as a `rrule(::typeof(f), ...)`
+No magic macro here, `rrule` is the function that it is.
+The function it is the rule for is the first argument, or second argument if you need to take a `[RuleConfig`](@ref).
+
+Note that when writing the rule for constructor you will need to use `::Type{Foo}`, not `typeof(Foo)`.
+See docs on [Constructors](@ref).
+
+## Include the derivative with respect to the function object itself
+The `ZygoteRules.@adjoint` macro automagically[^1] inserts a extra `nothing` in the return for the function it generates to represent the derivative of output with respect to the function object.
+ChainRules as a philosophy avoids magic as much as possible, and thus require you to return it explicitly.
+If it is a plain function (like `typeof(sin)`), then the differential will be [`NoTangent`](@ref).
+
+
+[^1]: unless you write it in functor form (i.e. `@adjoint (f::MyType)(args...)=...`), in that case like for `rrule` you need to include it explictly.
+
+## Tangent Type changes
+ChainRules uses tangent types that must represent vector spaces (i.e. tangent spaces).
+They need to have things like `+` defined on them.
+ZygoteRules takes a more adhoc approach to this.
+
+### `nothing` becomes an `AbstractZero`
+ZygoteRules uses nothing to represent some sense of zero, in a primal type agnostic way.
+There are many senses of zero.
+ChainRules represents two of them, as subtypes of [`AbstractZero`](@ref).
+
+[`ZeroTangent`](@ref) for the case that there is no relationship between the primal output and the primal input.
+[`NoTangent`](@ref) for the case that conceptually the tangent space don't exist.
+e.g. what is the Tangent to a String or an index: those can't be perturbed.
+
+See [FAQ on difference between `ZeroTangent` and `NoTangent`](@ref faq_abstract_zero).
+At the end of the day it doesn't matter too much if you get them wrong.
+`NoTangent` and `ZeroTangent` more or less act identically.
```

Comment:
ChainRulesTestUtils will complain if `ZeroTangent` is used instead of `NoTangent` for index-like arguments. Should we mention that, or just remove these two sentences?

Reply:
I think that this is covered by "doesn't matter *too* much"
and "*more or less* act identically"

We are not saying they are literally identical, or that it doesn't matter at all.

Adding more detail would i think detract from the point,
and ChainRulesTestUtils should give mostly good messages telling people how to fix it.

So I am fine with as is